File: info.jl
mutable struct Contact <: Comparable
    name::String
    url::URI
    email::String
end

Contact() = Contact("", URI(), "")

function parse!(c::Contact, data::Dict{Any,Any})
    for (key, value) in data
        if key == "name" c.name = value end
        if key == "url" c.url = URI(value) end
        if key == "email" c.email = value end
    end
end

mutable struct License <: Comparable
    name::String
    identifier::String
    url::URI
end

License() = License("", "", URI())

function parse!(l::License, data::Dict{Any,Any})
    for (key, value) in data
        if key == "name" l.name = value end
        if key == "identifier" l.identifier = value end
        if key == "url" l.url = URI(value) end
    end
end

mutable struct Info <: Comparable
    title::String
    summary::String
    description::String
    termsOfService::URI
    contact::Contact
    license::License
    version::String
end

Info() = Info("", "", "", URI(), Contact(), License(), "")

function parse!(info::Info, data::Dict{Any,Any})
    for (key, value) in data
        if key == "title" info.title = value end
        if key == "summary" info.summary = value end
        if key == "description" info.description = convert(value) end
        if key == "termsOfService" info.termsOfService = URI(value) end
        if key == "contact" parse!(info.contact, value) end
        if key == "license" parse!(info.license, value) end
        if key == "version" info.version = value end
    end
end

File: src/parser/info.cpp
//
// Created by Rakesh on 19/09/2025.
//

#include "parser.hpp"
#include "model/info.hpp"

template <>
void spt::parser::parse( model::Info::Contact& m, c4::yml::ConstNodeRef node )
{
  for ( const auto& child : node.children() )
  {
    if ( child.key() == "name" ) child >> m.name;
    if ( child.key() == "email" ) child >> m.email;
    if ( child.key() == "url" ) m.url = parse<boost::urls::url>( child );
  }
}

template <>
void spt::parser::parse( model::Info::License& m, c4::yml::ConstNodeRef node )
{
  for ( const auto& child : node.children() )
  {
    if ( child.key() == "name" ) child >> m.name;
    if ( child.key() == "identifier" ) child >> m.identifier;
    if ( child.key() == "url" ) m.url = parse<boost::urls::url>( child );
  }
}

template <>
void spt::parser::parse( model::Info& m, c4::yml::ConstNodeRef node )
{
  for ( const auto& child : node.children() )
  {
    if ( child.key() == "title" ) child >> m.title;
    if ( child.key() == "summary" ) child >> m.summary;
    if ( child.key() == "description" ) child >> m.description;
    if ( child.key() == "version" ) child >> m.version;
    if ( child.key() == "termsOfService" ) m.termsOfService = parse<boost::urls::url>( child );
    if ( child.key() == "license" ) m.license = parse<model::Info::License>( child );
    if ( child.key() == "contact" ) m.contact = parse<model::Info::Contact>( child );
  }
}
